### PR TITLE
[FIX] #4301: Skip partners without access.

### LINKED ISSUE #4301: 8.0 : Cannot access form view of partners when only in "Contact Creation" group
Here are the steps to reproduce the bug on OpenERP v8:
1) put the "demo" user only in the following groups : HR / Employee + Contact Creation
2) by default, even if the user is in the group "Contact Creation", he will not have access to a partner menuitem : add such a menuitem.
3) login as demo user, select a partner to open it's form view, and you get the following error:

AccessError

Sorry, you are not allowed to access this document. Only users with the following access level are currently allowed to do that:
- Point of Sale/Manager
- Point of Sale/User
- Accounting / Payments
- Mass/User
- Purchases/User
- Accounting & Finance/Invoicing & Payments
- Sales/See Own Leads
- Accounting & Finance/Financial Manager
- Donation/User

Here is the backtrace:

(Document model  File "/home/alexis/new_boite/dev/odoo/openerp/osv/fields.py", line 1356, in get
    result = self._fnct(obj, cr, uid, ids, name, self._arg, context)
  File "/home/alexis/new_boite/dev/odoo/addons/account/partner.py", line 257, in _journal_item_count
    for partner_id in ids
  File "/home/alexis/new_boite/dev/odoo/addons/account/partner.py", line 257, in <dictcomp>
    for partner_id in ids
  File "/home/alexis/new_boite/dev/odoo/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/alexis/new_boite/dev/odoo/openerp/models.py", line 1626, in search_count
    res = self.search(cr, user, args, context=context, count=True)
  File "/home/alexis/new_boite/dev/odoo/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/alexis/new_boite/dev/barroux/barroux_account/account.py", line 66, in search
    context=context, count=count)
  File "/home/alexis/new_boite/dev/odoo/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/alexis/new_boite/dev/odoo/addons/account/account_move_line.py", line 758, in search
    return super(account_move_line, self).search(cr, uid, args, offset, limit, order, context, count)
  File "/home/alexis/new_boite/dev/odoo/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/alexis/new_boite/dev/odoo/openerp/models.py", line 1647, in search
    return self._search(cr, user, args, offset=offset, limit=limit, order=order, context=context, count=count)
  File "/home/alexis/new_boite/dev/odoo/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/alexis/new_boite/dev/odoo/openerp/models.py", line 4555, in _search
    self.check_access_rights(cr, access_rights_uid or user, 'read')
  File "/home/alexis/new_boite/dev/odoo/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File "/home/alexis/new_boite/dev/odoo/openerp/models.py", line 3417, in check_access_rights
    return self.pool.get('ir.model.access').check(cr, uid, self._name, operation, raise_exception)
  File "/home/alexis/new_boite/dev/odoo/openerp/api.py", line 237, in wrapper
    return old_api(self, *args, **kwargs)
  File "<string>", line 2, in check
  File "/home/alexis/new_boite/dev/odoo/openerp/tools/cache.py", line 71, in lookup
    value = d[key] = self.method(*args, **kwargs)
  File "/home/alexis/new_boite/dev/odoo/openerp/addons/base/ir/ir_model.py", line 797, in check
    raise openerp.exceptions.AccessError(msg % msg_params)
AccessError: ('AccessError', u'Sorry, you are not allowed to access this document. Only users with the following access level are currently allowed to do that:\n- Point of Sale/Manager\n\t- Point of Sale/User\n\t- Accounting / Payments\n\t- Mass/User\n\t- Purchases/User\n\t- Accounting & Finance/Invoicing & Payments\n\t- Sales/See Own Leads\n\t- Accounting & Finance/Financial Manager\n\t- Donation/User\n\n(Document model: account.move.line)'): account.move.line) 

Code analysis : in addons/account/partner.py, there are 2 function fields 'contracts_count' and 'journal_item_count' that are used to display a cute logo in the top right corner of the partner form view. The function that compute these function fields suppose that you have read access to account.move.line, which is not always the case !

### PR DESCRIPTION
**Pull request by kiorky**
This fixes #4301 and replaces #4302.

### CODE REVIEW

**Comment by kiorky** [comment]
My cla access is in odoo/odoo#6132

**Comment by mart-e** [comment]
Hello,

Is this still needed as #4301 is marked as fixed ?

**Comment by kiorky** [comment]
Yes, it is

**Comment by mart-e** [comment]
So could you give me a way to reproduce the bug please ?

**Comment by kiorky** [comment]
Well, it will be hard, as it is with a custom set of modules & odoo version...
But basically we use odoo8, and modules tied to payment (SEPA & co).

We triggered the bug while logging to the XMLRPC api via the python binding (OERPLib)

If you want to be more confident of how we deployed github, you can give a shot to https://github.com/makinacorpus/corpus-odoo/tree/master/.salt (we deploy odoo with saltstack).

**Comment by kiorky** [comment]
cc @chaby1

**Comment by kiorky** [comment]
an error in rebase, im pushing a fix in the few minutes

**Comment by kiorky** [comment]
pushed

File: addons/account/partner.py [hunk 1]
```diff
@@ -252,12 +254,21 @@ def _journal_item_count(self, cr, uid, ids, field_name, arg, context=None):
         MoveLine = self.pool('account.move.line')
         AnalyticAccount = self.pool('account.analytic.account')
         results = {}
-        for partner_id in ids:
-            results[partner_id] = {}
-            if 'contracts_count' in field_name:
-                results[partner_id]['contracts_count'] = AnalyticAccount.search_count(cr, uid, [('partner_id', '=', partner_id)], context=context)
-            if 'journal_item_count' in field_name:
-                results[partner_id]['journal_item_count'] = MoveLine.search_count(cr, uid, [('partner_id', '=', partner_id)], context=context)
+        for partner_id in partner_ids:
```
Comment (mart-e) [inline addons/account/partner.py]: undefined variable partner_ids

File: addons/account/partner.py [hunk 2]
```diff
@@ -252,12 +254,21 @@ def _journal_item_count(self, cr, uid, ids, field_name, arg, context=None):
...
+                        cr,uid, [('partner_id', '=', partner_id)], context=context)
+                ),
+                'journal_item_count': lambda: (
+                    MoveLine.search_count(
+                        cr, uid, [('partner_id', '=', partner_id)], context=context)
+                )
+            }.items():
+                try:
+                    results.setdefault(partner_id, {})[field] = getter()
+                except Exception:
```
Comment (mart-e) [inline addons/account/partner.py]: never catch `Exception`, be specific, is it an `AccessError` ?

File: addons/account_followup/account_followup.py [hunk 3]
```diff
@@ -134,22 +136,25 @@ def _get_latest(self, cr, uid, ids, names, arg, context=None, company_id=None):
...
-                    latest_level = aml.followup_line_id.id
-                if (aml.company_id == company) and (not latest_date or latest_date < aml.followup_date):
-                    latest_date = aml.followup_date
-                if (aml.company_id == company) and (aml.blocked == False) and (aml.followup_line_id != False and 
-                            (not latest_days_without_lit or latest_days_without_lit < aml.followup_line_id.delay)):
-                    latest_days_without_lit =  aml.followup_line_id.delay
-                    latest_level_without_lit = aml.followup_line_id.id
+            try:
+                amls = partner.unreconciled_aml_ids
+                for aml in amls:
```
Comment (mart-e) [inline addons/account_followup/account_followup.py]: `for aml in partner.unreconciled_aml_ids:`

File: addons/purchase/partner.py [hunk 4]
```diff
@@ -28,23 +30,30 @@ class res_partner(osv.osv):
...
-                'purchase_order_count': PurchaseOrder.search_count(cr,uid, [('partner_id', 'child_of', partner_id)], context=context),
-                'supplier_invoice_count': Invoice.search_count(cr,uid, [('partner_id', 'child_of', partner_id), ('type','=','in_invoice')], context=context)
-            }
-            for partner_id in ids
-        }
+        results = {}
+        for partner_id in ids:
+            results[partner_id] = {}
+            try:
+                if 'contracts_count' in field_name:
```
Comment (mart-e) [inline addons/purchase/partner.py]: what is `contracts_count` doing in purchase compute method ?

File: addons/purchase/partner.py [hunk 5]
```diff
@@ -28,23 +30,30 @@ class res_partner(osv.osv):
...
+        return results
 
     def _commercial_fields(self, cr, uid, context=None):
         return super(res_partner, self)._commercial_fields(cr, uid, context=context) + ['property_product_pricelist_purchase']
 
     _columns = {
         'property_product_pricelist_purchase': fields.property(
-          type='many2one', 
-          relation='product.pricelist', 
+          type='many2one',
```
Comment (mart-e) [inline addons/purchase/partner.py]: to keep the diff minimal, avoid changing trailing spaces, it makes it harder to read the diff of the commit

**Comment by mart-e** [comment]
Thank you for the update.

I have put inline comments for corrections of your code. However I am not confident on the issue itself. If the `_journal_item_count` receives ids on records it has no access to, it seems normal for me to get an error, it's like this everywhere in odoo, not jsut on partner.

Now, if there is part of the code where you can not access some objects (e.g. some fields in `_get_latest`, but please specify which one), it may make sense to use superuser to access these.

However, without a valid scenario to reproduce the issue, I can not accept adding try/except everywhere.

File: addons/account/partner.py (hunk 1 above)
Comment (kiorky) [inline addons/account/partner.py]: rebase error, will update

File: addons/account/partner.py (hunk 2 above)
Comment (kiorky) [inline addons/account/partner.py]: lost with the rebase, will update( if you looked on the former changesets, yes im catching accesserror)

File: addons/purchase/partner.py (hunk 5 above)
Comment (kiorky) [inline addons/purchase/partner.py]: The former error is the one who commited whitespaces.

**Comment by kiorky** [comment]
The errors comes from some combination of addons, and after that API login is impossible.
As i said, to reproduce, mimic the salt deployment i gave you.

**Comment by kiorky** [comment]
(and then try to login via OERPlib with a non superuser)

**Comment by mart-e** [comment]
Sorry but I need more specific information than "try our code", I don't want to sound rude but I don't have the time to try the custom code/deployment of everybody. If you have a technical support contract, we can help you to debug your code but if you can not reproduce in standard or demonstrate easily a valid scenario where there is a bug, the source of the issue is probably your code, not the standard.

**Comment by kiorky** [comment]
As i said, the scenario is pretty simple, install odoo + the module i use, and login via OERPLib with a non super user, this wont work because of accesserror triggered during login.

**Comment by kiorky** [comment]
@mart-e PTAL, review concerns have been addressed.

**Comment by mart-e** [comment]
Again, I can not accept this PR like this. When I asked you to give us more information on the issue, I asked for an analysis of the issue like "displaying the list of all partners will execute X then Y, producing the error because..." not "install our module and figure out why it crashed". Moreover, I have no idea how to use saltstack and don't really want to start using external tool to reproduce a bug.

Your patch is catching silently errors, it's like hiding the symptoms but not curing the cause (and that's quite bad for maintainability). It may be correct but it's only if we really don't have other choice.
Here I have no idea (not even a traceback) what is causing the access errors you are facing so can not blindly say that this is the right fix.

I am closing this but if you can demonstrate me I am wrong, we can take another look at it.

**Comment by kiorky** [comment]
There are nemerous bugs linked to this PR which are not only from me ....

**Pull request closed without merging**